cis194/week4/tien/homework4/src/Homework4.hs

```diff
@@ -30,3 +30,34 @@ data Tree a
   = Leaf
   | Node Integer (Tree a) a (Tree a)
   deriving (Show, Eq)
+
+checkDepth :: Eq a => Tree a -> Integer
+checkDepth Leaf = (0 :: Integer)
+checkDepth (Node _ Leaf _ Leaf) = 1
+checkDepth (Node _ left _ right)
+  | left /= Leaf && right == Leaf = 1 + checkDepth left
+  | right /= Leaf && left == Leaf = 1 + checkDepth right
+  | otherwise = 1 + max (checkDepth left) (checkDepth right)
+
+insert :: Eq t => t -> Tree t -> Tree t
+insert x Leaf = Node 0 Leaf x Leaf
+insert x (Node i left c Leaf) = Node i left c (Node (i + 1) Leaf x Leaf)
+insert x (Node i Leaf c right) = Node i (Node (i + 1) Leaf x Leaf) c right
+insert x (Node i left c right)
+  | (checkDepth left) > (checkDepth right) = Node i left c (insert x right)
+  | otherwise = Node i (insert x left) c right
+
+foldTree :: (Foldable t, Eq a) => t a -> Tree a
+foldTree arr
+  | length arr == 0 = Leaf
+  | otherwise = foldr insert Leaf arr
+
+-- Exercise 3.1 --
+xor :: [Bool] -> Bool
+xor = odd . foldr (\x acc -> acc + (bool 0 1 x)) (0 :: Integer)
+-- Exercise 3.2 --
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\x acc-> f x:acc) []
+-- Exercise 3.3 --
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f base xs = foldr (\x acc -> acc `f` x) base xs
```

Comment:
is `foldl (flip (:)) [] [1, 2, 3]` the same as `myFoldl (flip (:)) [] [1, 2, 3]`?

Reply:
@shineli1984 I cannot see any difference between the two.

Reply:
@cht8687 because (-) is used in test for this function, we can not detect any difference because the arguments for (-) are both Num.

Reply:
Fixed.